File: luaapp1/game.lua
-- 推箱子核心逻辑（Lua 教学）

local M = {}

local function key(x, y)
  return x .. "," .. y
end

function M.from_rows(rows, index)
  local s = {
    walls = {},
    goals = {},
    boxes = {},
    player = { x = 0, y = 0 },
    moves = 0,
    won = false,
    width = 0,
    height = 0,
    level_index = index or 0,
    hist = {},
  }
  local max_x, max_y = 0, 0
  for y, row in ipairs(rows) do
    y = y - 1 -- 0-based
    max_y = y
    for x = 1, #row do
      local ch = row:sub(x, x)
      local xi = x - 1
      if xi > max_x then max_x = xi end
      local k = key(xi, y)
      if ch == "#" then
        s.walls[k] = true
      elseif ch == "." then
        s.goals[k] = true
      elseif ch == "$" then
        s.boxes[k] = true
      elseif ch == "*" then
        s.boxes[k] = true
        s.goals[k] = true
      elseif ch == "@" then
        s.player = { x = xi, y = y }
      elseif ch == "+" then
        s.player = { x = xi, y = y }
        s.goals[k] = true
      end
    end
  end
  s.width = max_x + 1
  s.height = max_y + 1
  return s
end

local function check_win(s)
  for b, _ in pairs(s.boxes) do
    if not s.goals[b] then
      s.won = false
      return
    end
  end
  s.won = true
end

function M.try_move(s, dx, dy)
  if s.won then return false end
  local nx = s.player.x + dx
  local ny = s.player.y + dy
  local nk = key(nx, ny)
  if s.walls[nk] then return false end
  if s.boxes[nk] then
    local bx = nx + dx
    local by = ny + dy
    local bk = key(bx, by)
    if s.walls[bk] or s.boxes[bk] then return false end
    s.hist[#s.hist + 1] = {
      player = { x = s.player.x, y = s.player.y },
      box_from = nk,
      box_to = bk,
    }
    s.boxes[nk] = nil
    s.boxes[bk] = true
    s.player.x, s.player.y = nx, ny
    s.moves = s.moves + 1
    check_win(s)
    return true
  end
  s.hist[#s.hist + 1] = {
    player = { x = s.player.x, y = s.player.y },
  }
  s.player.x, s.player.y = nx, ny
  return true
end

function M.undo(s)
  if s.won or #s.hist == 0 then return false end
  local entry
  while #s.hist > 0 do
    entry = table.remove(s.hist)
    if entry.box_from then break end
    s.player = entry.player
  end
  if not entry or not entry.box_from then return true end
  s.player = entry.player
  s.boxes[entry.box_to] = nil
  s.boxes[entry.box_from] = true
  if s.moves > 0 then s.moves = s.moves - 1 end
  s.won = false
  return true
end

function M.render_ascii(s)
  local lines = {}
  for y = 0, s.height - 1 do
    local row = {}
    for x = 0, s.width - 1 do
      local k = key(x, y)
      local ch
      if s.player.x == x and s.player.y == y then
        ch = s.goals[k] and "+" or "@"
      elseif s.boxes[k] then
        ch = s.goals[k] and "*" or "$"
      elseif s.walls[k] then
        ch = "#"
      elseif s.goals[k] then
        ch = "."
      else
        ch = " "
      end
      row[#row + 1] = ch
    end
    lines[#lines + 1] = table.concat(row)
  end
  return table.concat(lines, "\n") .. "\n"
end

return M

File: rapp1/game.R
# 推箱子核心逻辑（R 教学）

key <- function(x, y) paste0(x, ",", y)

from_rows <- function(rows, index = 0L) {
  walls <- character(0)
  goals <- character(0)
  boxes <- character(0)
  player <- c(0L, 0L)
  max_x <- 0L
  max_y <- 0L
  for (y in seq_along(rows)) {
    yi <- as.integer(y - 1L)
    max_y <- yi
    row <- rows[[y]]
    chars <- strsplit(row, "", fixed = TRUE)[[1]]
    for (x in seq_along(chars)) {
      xi <- as.integer(x - 1L)
      if (xi > max_x) max_x <- xi
      ch <- chars[[x]]
      k <- key(xi, yi)
      if (ch == "#") walls <- c(walls, k)
      else if (ch == ".") goals <- c(goals, k)
      else if (ch == "$") boxes <- c(boxes, k)
      else if (ch == "*") {
        boxes <- c(boxes, k)
        goals <- c(goals, k)
      } else if (ch == "@") player <- c(xi, yi)
      else if (ch == "+") {
        player <- c(xi, yi)
        goals <- c(goals, k)
      }
    }
  }
  list(
    walls = unique(walls),
    goals = unique(goals),
    boxes = unique(boxes),
    player = player,
    moves = 0L,
    won = FALSE,
    width = max_x + 1L,
    height = max_y + 1L,
    level_index = as.integer(index),
    hist = list()
  )
}

check_win <- function(state) {
  all(state$boxes %in% state$goals)
}

try_move <- function(state, dx, dy) {
  if (isTRUE(state$won)) return(state)
  px <- state$player[1]
  py <- state$player[2]
  nx <- px + dx
  ny <- py + dy
  nk <- key(nx, ny)
  if (nk %in% state$walls) return(state)
  if (nk %in% state$boxes) {
    bx <- nx + dx
    by <- ny + dy
    bk <- key(bx, by)
    if (bk %in% state$walls || bk %in% state$boxes) return(state)
    state$hist[[length(state$hist) + 1L]] <- list(
      player = c(px, py), box_from = nk, box_to = bk
    )
    state$boxes <- setdiff(state$boxes, nk)
    state$boxes <- c(state$boxes, bk)
    state$player <- c(nx, ny)
    state$moves <- state$moves + 1L
    state$won <- check_win(state)
    return(state)
  }
  state$hist[[length(state$hist) + 1L]] <- list(
    player = c(px, py), box_from = NULL, box_to = NULL
  )
  state$player <- c(nx, ny)
  state
}

undo <- function(state) {
  if (isTRUE(state$won) || length(state$hist) == 0L) return(state)
  entry <- NULL
  while (length(state$hist) > 0L) {
    entry <- state$hist[[length(state$hist)]]
    state$hist <- state$hist[-length(state$hist)]
    if (!is.null(entry$box_from)) break
    state$player <- entry$player
  }
  if (is.null(entry) || is.null(entry$box_from)) return(state)
  state$player <- entry$player
  state$boxes <- setdiff(state$boxes, entry$box_to)
  state$boxes <- c(state$boxes, entry$box_from)
  if (state$moves > 0L) state$moves <- state$moves - 1L
  state$won <- FALSE
  state
}

render_ascii <- function(state) {
  lines <- character(state$height)
  for (y in seq_len(state$height) - 1L) {
    row <- character(state$width)
    for (x in seq_len(state$width) - 1L) {
      k <- key(x, y)
      if (state$player[1] == x && state$player[2] == y) {
        row[x + 1L] <- if (k %in% state$goals) "+" else "@"
      } else if (k %in% state$boxes) {
        row[x + 1L] <- if (k %in% state$goals) "*" else "$"
      } else if (k %in% state$walls) {
        row[x + 1L] <- "#"
      } else if (k %in% state$goals) {
        row[x + 1L] <- "."
      } else {
        row[x + 1L] <- " "
      }
    }
    lines[y + 1L] <- paste0(row, collapse = "")
  }
  paste0(paste(lines, collapse = "\n"), "\n")
}
